src/frontend/src/expr/utils.rs

```diff
@@ -498,15 +498,32 @@ impl WatermarkAnalyzer {
                 }
                 _ => WatermarkDerivation::None,
             },
-            ExprType::Subtract
+            ty @ (ExprType::Subtract
             | ExprType::Divide
             | ExprType::TumbleStart
-            | ExprType::AtTimeZone => match self.visit_binary_op(func_call.inputs()) {
+            | ExprType::AtTimeZone) => match self.visit_binary_op(func_call.inputs()) {
                 (WatermarkDerivation::Constant, WatermarkDerivation::Constant) => {
                     WatermarkDerivation::Constant
                 }
                 (WatermarkDerivation::Watermark(idx), WatermarkDerivation::Constant) => {
-                    WatermarkDerivation::Watermark(idx)
+                    if ty == ExprType::AtTimeZone
+                        && !(func_call.return_type() == DataType::Timestamptz
+                            && func_call.inputs()[0].return_type() == DataType::Timestamp)
```

Comment:
What about writing the case `func_call.return_type() == DataType::Timestamp && func_call.inputs()[0].return_type() == DataType::Timestamptz` directly without `!`?

Reply:
> What about writing the case `func_call.return_type() == DataType::Timestamp && func_call.inputs()[0].return_type() == DataType::Timestamptz` directly without `!`?

I was afraid that in the future we might support conversion from one timestamp to another timestamp.